articles/kaneshin.re

```diff
@@ -0,0 +1,413 @@
+var ErrUnexpectedEOF = errors.New("unexpected EOF")
+var ErrNoProgress = errors.New("multiple Read calls return no data or error")
+//}
+
+パッケージスコープのエラーは定式的に@<code>{ErrNoProgress}のように@<code>{Err}をプレフィックスとすることが多いです。
+
+このようにハンドリングする必要のあるエラーは関数内で生成して返却するのではなく、パッケージスコープとして宣言したものを返却することでパッケージの外からハンドリングすることができます。
+
+=== 高コストな処理を初期化する
+
+Goの@<code>{regexp}パッケージは正規表現を扱うパッケージです。常に改善されているパッケージですが、関数内で何度も変数として初期化される場合は初期化のコストを実行直後に寄せてしまうことも可能です。
+
+//list[regexp_benchmark][regexpパッケージのベンチマーク][go]{
+package main
+
+import "regexp"
+import "testing"
+
+func BenchmarkRegexpMustCompile1(b *testing.B) {
+  b.ResetTimer()
+  for i := 0; i < b.N; i++ {
+    re0 := regexp.MustCompile("[a-z]{3}")
+    _ = re0.FindAllString(`Lorem ipsum dolor sit amet, consectetur..`, -1)
+  }
+}
+
+var re = regexp.MustCompile("[a-z]{3}")
+
+func BenchmarkRegexpMustCompile2(b *testing.B) {
+  b.ResetTimer()
+  for i := 0; i < b.N; i++ {
+    _ = re.FindAllString(`Lorem ipsum dolor sit amet, consectetur..`, -1)
+  }
+}
+//}
+
+//cmd{
+$ go test -bench . -benchmem
+BenchmarkRegexpMustCompile1-16 495222 2413 ns/op 1618 B/op 23 allocs/op
+BenchmarkRegexpMustCompile2-16 838984 1391 ns/op  289 B/op  9 allocs/op
+//}
+
+これは@<code>{regexp}を関数呼び出し時に生成するか、実行時に変数を初期化してしまうかをシミュレートしたベンチマークです。結果は実行時に初期化する方が2倍高速に処理ができます。このように@<code>{regexp}は正規表現が動的に変化しない場合はパッケージスコープの変数として宣言してしまうことで効率的に処理ができます。
+
+@<code>{regexp}は内部処理ではスレッドセーフに設計されていることや検索結果をキャッシュしているため、実際にコーディングする場合は次のように記述することを推奨します。
+
+//list[regexp_practice][regexpのプラクティス][go]{
+package main
+
+import "fmt"
+import "regexp"
+
+var re = regexp.MustCompile("[a-z]{3}")
+
+func FindString(str string) []string {
+  re := re.Copy()
+  return re.FindAllString(str, -1)
+}
+
+func main() {
+  fmt.Println(FindString("Hello, world"))
+}
+//}
+
+//cmd{
+$ go run main.
+[ell wor]
+//}
+
+また、パッケージ変数として宣言している@<code>{re}変数を上書きされたくない場合は初期化時に関数実行をして隠蔽することもできます。
+
+//list[regexp_practice_closure][regexpのプラクティス][go]{
+var FindString = func() func(string) []string {
+  re := regexp.MustCompile("[a-z]{3}")
+  return func(str string) []string {
+    re := re.Copy()
+    return re.FindAllString(str, -1)
+  }
+}()
+
+func main() {
+  fmt.Println(FindString("Hello, world"))
+}
+//}
+
+//cmd{
+$ go run main.
+[ell wor]
+//}
+
+このように宣言することによって、パッケージスコープの変数を減らすことができます。また、@<code>{re}変数が関数スコープになるためメモリアクセスも効率化されています。
+
+=== その他
+
+他にもさまざまな条件下で有効に活用されるパッケージスコープの変数が存在します。もちろん、ここに紹介していることに限らないのでメモリ効率のためには臆せず使うことも考えてみてください。
+
+==== メモリキャッシュ
+
+メモリキャッシュをするときもパッケージスコープに@<code>{map}や@<code>{map}をメンバ変数とした@<code>{struct}を変数としてキャッシュにすることもあります。並行処理で使われることが前提になるため競合状態のハンドリングを忘れずに実装する必要があります。@<code>{sync.Map}などを参考にしてみてください。
+
+==== 読み取り専用のテーブル
+
+@<code>{unicode}パッケージにはさまざまな文字に対応するためのUnicodeテーブル@<fn>{unicode_tables}が定義されています。
+//footnote[unicode_tables][@<href>{https://golang.org/src/unicode/tables.go}]
+
+== おわりに
+
+Goにおける初期化について、パッケージスコープの変数と@<code>{init}関数、初期化の順序について紹介しました。これらを気にすることなくコーディングすることは可能ですが、細かいところまでケアしてコードを書くことによってメモリや処理の最適化がされますし、開発における設計の複雑性除去の判断にもつながると筆者は考えています。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Line 411 sentence length(104) exceeds the maximum sentence length of 100.
Over 4 characters. (ja-technical-writing/sentence-length)